cmd/infra/aws/destroy.go

```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
+	// Clean up the OIDC S3 bucket so it can be deleted along with the stack
+	bucket := getStackOutput(stack, "OIDCBucketName")
+	err = emptyBucket(ctx, bucket, s3)
+	if err != nil {
+		return fmt.Errorf("failed to empty the OIDC bucket: %w", err)
+	}
+	log.Info("Emptied OIDC bucket", "id", bucket)
+
+	// Delete the NS record from the base domain hosted zone
+	baseDomainZoneID := getStackOutput(stack, "BaseDomainHostedZoneId")
+	subdomain := getStackOutput(stack, "Subdomain")
+	err = deleteRecord(ctx, r53, baseDomainZoneID, "NS", subdomain)
+	if err != nil {
+		return fmt.Errorf("failed to clean up the base domain zone: %w", err)
+	}
+	log.Info("Cleaned up the base domain hosted zone", "id", baseDomainZoneID, "subdomain", subdomain)
+
+	// Find and delete any non-default unmanaged DNS records
+	subdomainPrivateZoneID := getStackOutput(stack, "SubdomainPrivateZoneId")
+	err = deleteNonDefaultRecords(ctx, r53, subdomainPrivateZoneID)
+	if err != nil {
+		return fmt.Errorf("failed to clean up the subdomain private zone: %w", err)
+	}
+	log.Info("Cleaned up the subdomain private zone", "zone", subdomainPrivateZoneID)
+	subdomainPublicZoneID := getStackOutput(stack, "SubdomainPublicZoneId")
+	err = deleteNonDefaultRecords(ctx, r53, subdomainPublicZoneID)
+	if err != nil {
+		return fmt.Errorf("failed to clean up the subdomain public zone: %w", err)
+	}
+	log.Info("Cleaned up the subdomain public zone", "zone", subdomainPublicZoneID)
+
+	vpcID := getStackOutput(stack, "VPCId")
+
+	// Find and delete any unmanaged leaked load balancers
+	err = deleteUnmanagedELBs(ctx, elbclient, vpcID)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to delete unmanaged ELBs: %w", err)
 	}
-	route53client, err := route53Client(o.AWSCredentialsFile)
+	log.Info("Cleaned up unmanaged ELBs")
+
+	// Find and delete any unmanaged leaked security groups
+	err = deleteUnmanagedSecurityGroups(ctx, ec2client, vpcID)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to delete unmanaged security groups: %w", err)
 	}
-	elbclient, err := elbClient(o.AWSCredentialsFile, o.Region)
+	log.Info("Cleaned up unmanaged security groups")
+
+	// Delete the stack itself
+	err = deleteStack(ctx, cf, stack)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to delete stack: %w", err)
 	}
-	errs = append(errs, o.DestroyInternetGateways(ctx, ec2client)...)
-	errs = append(errs, o.DestroyVPCs(ctx, ec2client, elbclient)...)
-	errs = append(errs, o.DestroyDHCPOptions(ctx, ec2client)...)
-	errs = append(errs, o.DestroyEIPs(ctx, ec2client)...)
-	errs = append(errs, o.DestroyDNS(ctx, route53client)...)
-	return utilerrors.NewAggregate(errs)
+	log.Info("Deleted the stack", "id", *stack.StackId)
+
+	return nil
 }
 
-func (o *DestroyInfraOptions) DestroyELBs(ctx context.Context, client elbiface.ELBAPI, vpcID *string) []error {
+func deleteUnmanagedELBs(ctx context.Context, client elbiface.ELBAPI, vpcID string) error {
```

Comment:
Should have a TODO to delete NLBs as well. Also, I wonder if volumes need to be explicitly cleaned up as well.

Reply:
Yeah, unfortunately we're wrestling with the following facts now:

1. k8s (and users?) can create arbitrary resources that will become dependencies of Stack resources
2. Stack resources can only be deleted by AWS

So when deleting the Stack given some unbounded/unknown future set of dependent resources associated with the Stack resources, we have a couple of options I know of:

1. Do what this PR does which is try to identify any non-stack resources and destroy them so that Stack deletion will succeed
2. Use the Stack deletion setting which deletes the Stack regardless of dependencies, orphaning the non-Stack resources

Not a great situation.

Reply:
I do agree NLB and some volume stuff is inevitable to cover the stuff we _know_ OCP will do by default, but apparently we're not yet exercising any of those features in our e2e clusters.